File: AWS_GUI_Dev/BAWGUI.Settings/Model/BaseClassOrSharedClass.vb
Imports System.Collections.ObjectModel
Imports BAWGUI.Core
Imports BAWGUI.Settings.ViewModel
Imports BAWGUI.Settings.ViewModels

Namespace Model
    Public Class ParameterValuePair
        Inherits ViewModelBase
        Public Sub New()
            _isRequired = True
        End Sub
        Public Sub New(para As String, value As Object, required As Boolean)
            _parameterName = para
            _value = value
            _isRequired = required
        End Sub
        Public Sub New(para As String, value As Object)
            _parameterName = para
            _value = value
            _isRequired = True
        End Sub
        Private _parameterName As String
        Public Property ParameterName As String
            Get
                Return _parameterName
            End Get
            Set(ByVal value As String)
                _parameterName = value
                OnPropertyChanged()
            End Set
        End Property
        Private _value As Object
        Public Property Value As Object
            Get
                Return _value
            End Get
            Set(ByVal value As Object)
                _value = value
                OnPropertyChanged()
            End Set
        End Property
        Private _isRequired As Boolean
        Public Property IsRequired As Boolean
            Get
                Return _isRequired
            End Get
            Set(ByVal value As Boolean)
                _isRequired = value
                OnPropertyChanged()
            End Set
        End Property
        Private _toolTip As String
        Public Property ToolTip As String
            Get
                Return _toolTip
            End Get
            Set(ByVal value As String)
                _toolTip = value
                OnPropertyChanged()
            End Set
        End Property
    End Class

    ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    '''''''''''''''''''''''''''''''''Class PMU''''''''''''''''''''''''''''''''''''''''''''''''''''''
    ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    'Public Class PMU
    '    Inherits ViewModelBase
    '    Public Sub New()
    '        _channels = New ObservableCollection(Of String)
    '    End Sub
    '    Private _PMUName As String
    '    Public Property PMUName As String
    '        Get
    '            Return _PMUName
    '        End Get
    '        Set(ByVal value As String)
    '            _PMUName = value
    '            OnPropertyChanged()
    '        End Set
    '    End Property
    '    Private _channels As ObservableCollection(Of String)
    '    Public Property Channels As ObservableCollection(Of String)
    '        Get
    '            Return _channels
    '        End Get
    '        Set(ByVal value As ObservableCollection(Of String))
    '            _channels = value
    '            OnPropertyChanged()
    '        End Set
    '    End Property
    'End Class

    '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    ''''''''''''''''''''''''''''''''Class Filter''''''''''''''''''''''''''''''''''''''''''''''''''''''
    '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

    '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    ''''''''''''''''''''''''''''''''Class SignalProcessStep''''''''''''''''''''''''''''''''''''''''
    '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    ''' <summary>
    ''' This is the base class that all signal process steps inherits, including
    ''' filters (DQFilter and TunableFilter), data config customization, unwrap, wrap, interpolate, etc.
    ''' </summary>
    Public MustInherit Class SignalProcessStep
        Inherits ViewModelBase

        Public MustOverride Property Name As String
        Private _stepCounter As Integer
        Public Property StepCounter As Integer
            Get
                Return _stepCounter
            End Get
            Set(ByVal value As Integer)
                _stepCounter = value

                OnPropertyChanged()
            End Set
        End Property

        Private _isStepSelected As Boolean
        Public Property IsStepSelected As Boolean
            Get
                Return _isStepSelected
            End Get
            Set(ByVal value As Boolean)
                _isStepSelected = value
                OnPropertyChanged()
            End Set
        End Property

        Private _inputChannels As ObservableCollection(Of SignalSignatureViewModel)
        Public Property InputChannels As ObservableCollection(Of SignalSignatureViewModel)
            Get
                Return _inputChannels
            End Get
            Set(ByVal value As ObservableCollection(Of SignalSignatureViewModel))
                _inputChannels = value
                OnPropertyChanged()
            End Set
        End Property

        Private _thisStepInputsAsSignalHierachyByType As SignalTypeHierachy
        Public Property ThisStepInputsAsSignalHerachyByType As SignalTypeHierachy
            Get
                Return _thisStepInputsAsSignalHierachyByType
            End Get
            Set(ByVal value As SignalTypeHierachy)
                _thisStepInputsAsSignalHierachyByType = value
                OnPropertyChanged()
            End Set
        End Property

        Private _outputChannels As ObservableCollection(Of SignalSignatureViewModel)
        Public Property OutputChannels As ObservableCollection(Of SignalSignatureViewModel)
            Get
                Return _outputChannels
            End Get
            Set(value As ObservableCollection(Of SignalSignatureViewModel))
                _outputChannels = value
                OnPropertyChanged()
            End Set
        End Property

        Private _thisStepOutputsAsSignalHierachyByPMU As SignalTypeHierachy
        Public Property ThisStepOutputsAsSignalHierachyByPMU As SignalTypeHierachy
            Get
                Return _thisStepOutputsAsSignalHierachyByPMU
            End Get
            Set(ByVal value As SignalTypeHierachy)
                _thisStepOutputsAsSignalHierachyByPMU = value
                OnPropertyChanged()
            End Set
        End Property

        Private _isExpanded As Boolean
        Public Property IsExpanded As Boolean
            Get
                Return _isExpanded
            End Get
            Set(ByVal value As Boolean)
                _isExpanded = value
                OnPropertyChanged()
            End Set
        End Property

        Public MustOverride Function CheckStepIsComplete() As Boolean
    End Class
    Public Class Filter
        Inherits SignalProcessStep
        Public Sub New()
            '_filterParameters = New ObservableCollection(Of ParameterValuePair)
            _fileterParameters = New ObservableCollection(Of ParameterValuePair)
            InputChannels = New ObservableCollection(Of SignalSignatureViewModel)
            ThisStepInputsAsSignalHerachyByType = New SignalTypeHierachy(New SignalSignatureViewModel)

            OutputChannels = New ObservableCollection(Of SignalSignatureViewModel)
            ThisStepOutputsAsSignalHierachyByPMU = New SignalTypeHierachy(New SignalSignatureViewModel)
            IsExpanded = False
        End Sub

        Public Overrides Function CheckStepIsComplete() As Boolean
            Return True
        End Function

        Private _name As String
        Public Overrides Property Name As String
            Get
                Return _name
            End Get
            Set(value As String)
                _name = value
                OnPropertyChanged()
            End Set
        End Property

        'Private _filterName As String
        'Public Property FilterName As String
        '    Get
        '        Return _filterName
        '    End Get
        '    Set(ByVal value As String)
        '        _filterName = value
        '        OnPropertyChanged()
        '    End Set
        'End Property

        Private _fileterParameters As ObservableCollection(Of ParameterValuePair)
        Public Property FilterParameters As ObservableCollection(Of ParameterValuePair)
            Get
                Return _fileterParameters
            End Get
            Set(ByVal value As ObservableCollection(Of ParameterValuePair))
                _fileterParameters = value
                OnPropertyChanged()
            End Set
        End Property

        'Private _filterParameters As ObservableCollection(Of ParameterValuePair)
        'Public Property FilterParameters As ObservableCollection(Of ParameterValuePair)
        '    Get
        '        Return _filterParameters
        '    End Get
        '    Set(ByVal value As ObservableCollection(Of ParameterValuePair))
        '        _filterParameters = value
        '        OnPropertyChanged()
        '    End Set
        'End Property
    End Class

    Public MustInherit Class DetectorBase
        Inherits ViewModelBase

        Public MustOverride ReadOnly Property Name As String

        Private _isStepSelected As Boolean
        Public Property IsStepSelected As Boolean
            Get
                Return _isStepSelected
            End Get
            Set(ByVal value As Boolean)
                _isStepSelected = value
                OnPropertyChanged()
            End Set
        End Property

        Private _inputChannels As ObservableCollection(Of SignalSignatureViewModel)
        Public Property InputChannels As ObservableCollection(Of SignalSignatureViewModel)
            Get
                Return _inputChannels
            End Get
            Set(ByVal value As ObservableCollection(Of SignalSignatureViewModel))
                _inputChannels = value
                OnPropertyChanged()
            End Set
        End Property
        Private _thisStepInputsAsSignalHierachyByType As SignalTypeHierachy
        Public Property ThisStepInputsAsSignalHerachyByType As SignalTypeHierachy
            Get
                Return _thisStepInputsAsSignalHierachyByType
            End Get
            Set(ByVal value As SignalTypeHierachy)
                _thisStepInputsAsSignalHierachyByType = value
                OnPropertyChanged()
            End Set
        End Property
        Private _isExpanded As Boolean
        Public Property IsExpanded As Boolean
            Get
                Return _isExpanded
            End Get
            Set(ByVal value As Boolean)
                _isExpanded = value
                OnPropertyChanged()
            End Set
        End Property
    End Class

    Public MustInherit Class AlarmingDetectorBase
        Inherits ViewModelBase

        Public MustOverride ReadOnly Property Name As String

        Private _isStepSelected As Boolean
        Public Property IsStepSelected As Boolean
            Get
                Return _isStepSelected
            End Get
            Set(ByVal value As Boolean)
                _isStepSelected = value
                OnPropertyChanged()
            End Set
        End Property
        Private _isExpanded As Boolean
        Public Property IsExpanded As Boolean
            Get
                Return _isExpanded
            End Get
            Set(ByVal value As Boolean)
                _isExpanded = value
                OnPropertyChanged()
            End Set
        End Property
    End Class
End Namespace

File: AWS_GUI_Dev/BAWGUI.Core/ViewModels/BaseClassOrSharedClass.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System.Collections.ObjectModel;
using BAWGUI.Utilities;
using BAWGUI.SignalManagement.ViewModels;

namespace BAWGUI.Core
{
    //public class ParameterValuePair : ViewModelBase
    //{
    //    public ParameterValuePair()
    //    {
    //        _isRequired = true;
    //    }
    //    public ParameterValuePair(string para, object value, bool required)
    //    {
    //        _parameterName = para;
    //        _value = value;
    //        _isRequired = required;
    //    }
    //    public ParameterValuePair(string para, object value)
    //    {
    //        _parameterName = para;
    //        _value = value;
    //        _isRequired = true;
    //    }
    //    private string _parameterName;
    //    public string ParameterName
    //    {
    //        get
    //        {
    //            return _parameterName;
    //        }
    //        set
    //        {
    //            _parameterName = value;
    //            OnPropertyChanged();
    //        }
    //    }
    //    private object _value;
    //    public object Value
    //    {
    //        get
    //        {
    //            return _value;
    //        }
    //        set
    //        {
    //            _value = value;
    //            OnPropertyChanged();
    //        }
    //    }
    //    private bool _isRequired;
    //    public bool IsRequired
    //    {
    //        get
    //        {
    //            return _isRequired;
    //        }
    //        set
    //        {
    //            _isRequired = value;
    //            OnPropertyChanged();
    //        }
    //    }
    //    private string _toolTip;
    //    public string ToolTip
    //    {
    //        get
    //        {
    //            return _toolTip;
    //        }
    //        set
    //        {
    //            _toolTip = value;
    //            OnPropertyChanged();
    //        }
    //    }
    //}

     //''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
     //'''''''''''''''''''''''''''''''Class SignalProcessStep''''''''''''''''''''''''''''''''''''''''
     //''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    public abstract class SignalProcessStep : ViewModelBase
    {
        private int _stepCounter;
        public int StepCounter
        {
            get
            {
                return _stepCounter;
            }
            set
            {
                _stepCounter = value;

                OnPropertyChanged();
            }
        }

        private bool _isStepSelected;
        public bool IsStepSelected
        {
            get
            {
                return _isStepSelected;
            }
            set
            {
                _isStepSelected = value;
                OnPropertyChanged();
            }
        }

        private ObservableCollection<SignalSignatureViewModel> _inputChannels;
        public ObservableCollection<SignalSignatureViewModel> InputChannels
        {
            get
            {
                return _inputChannels;
            }
            set
            {
                _inputChannels = value;
                OnPropertyChanged();
            }
        }

        private SignalTypeHierachy _thisStepInputsAsSignalHierachyByType;
        public SignalTypeHierachy ThisStepInputsAsSignalHerachyByType
        {
            get
            {
                return _thisStepInputsAsSignalHierachyByType;
            }
            set
            {
                _thisStepInputsAsSignalHierachyByType = value;
                OnPropertyChanged();
            }
        }

        private ObservableCollection<SignalSignatureViewModel> _outputChannels;
        public ObservableCollection<SignalSignatureViewModel> OutputChannels
        {
            get
            {
                return _outputChannels;
            }
            set
            {
                _outputChannels = value;
                OnPropertyChanged();
            }
        }

        private SignalTypeHierachy _thisStepOutputsAsSignalHierachyByPMU;
        public SignalTypeHierachy ThisStepOutputsAsSignalHierachyByPMU
        {
            get
            {
                return _thisStepOutputsAsSignalHierachyByPMU;
            }
            set
            {
                _thisStepOutputsAsSignalHierachyByPMU = value;
                OnPropertyChanged();
            }
        }

        private bool _isExpanded;
        public bool IsExpanded
        {
            get
            {
                return _isExpanded;
            }
            set
            {
                _isExpanded = value;
                OnPropertyChanged();
            }
        }

        public abstract bool CheckStepIsComplete();
    }
    public class Filter : SignalProcessStep
    {
        public Filter()
        {
            //_fileterParameters = new ObservableCollection<ParameterValuePair>();
            InputChannels = new ObservableCollection<SignalSignatureViewModel>();
            ThisStepInputsAsSignalHerachyByType = new SignalTypeHierachy(new SignalSignatureViewModel());

            OutputChannels = new ObservableCollection<SignalSignatureViewModel>();
            ThisStepOutputsAsSignalHierachyByPMU = new SignalTypeHierachy(new SignalSignatureViewModel());
            IsExpanded = false;
        }

        public override bool CheckStepIsComplete()
        {
            return true;
        }

        private string _name;
        public string Name
        {
            get
            {
                return _name;
            }
            set
            {
                _name = value;
                OnPropertyChanged();
            }
        }

        // Private _filterName As String
        // Public Property FilterName As String
        // Get
        // Return _filterName
        // End Get
        // Set(ByVal value As String)
        // _filterName = value
        // OnPropertyChanged()
        // End Set
        // End Property

        //private ObservableCollection<ParameterValuePair> _fileterParameters;
        //public ObservableCollection<ParameterValuePair> FilterParameters
        //{
        //    get
        //    {
        //        return _fileterParameters;
        //    }
        //    set
        //    {
        //        _fileterParameters = value;
        //        OnPropertyChanged();
        //    }
        //}
    }

    public abstract class DetectorBase : ViewModelBase
    {
        public abstract string Name { get; }

        private bool _isStepSelected;
        public bool IsStepSelected
        {
            get
            {
                return _isStepSelected;
            }
            set
            {
                _isStepSelected = value;
                OnPropertyChanged();
            }
        }

        private ObservableCollection<SignalSignatureViewModel> _inputChannels;
        public ObservableCollection<SignalSignatureViewModel> InputChannels
        {
            get
            {
                return _inputChannels;
            }
            set
            {
                _inputChannels = value;
                OnPropertyChanged();
            }
        }
        private SignalTypeHierachy _thisStepInputsAsSignalHierachyByType;
        public SignalTypeHierachy ThisStepInputsAsSignalHerachyByType
        {
            get
            {
                return _thisStepInputsAsSignalHierachyByType;
            }
            set
            {
                _thisStepInputsAsSignalHierachyByType = value;
                OnPropertyChanged();
            }
        }
        private bool _isExpanded;
        public bool IsExpanded
        {
            get
            {
                return _isExpanded;
            }
            set
            {
                _isExpanded = value;
                OnPropertyChanged();
            }
        }
    }

    public abstract class AlarmingDetectorBase : ViewModelBase
    {
        public abstract string Name { get; }

        private bool _isStepSelected;
        public bool IsStepSelected
        {
            get
            {
                return _isStepSelected;
            }
            set
            {
                _isStepSelected = value;
                OnPropertyChanged();
            }
        }
        private bool _isExpanded;
        public bool IsExpanded
        {
            get
            {
                return _isExpanded;
            }
            set
            {
                _isExpanded = value;
                OnPropertyChanged();
            }
        }
    }
}
